### ISSUE TITLE
[cache tagging] add tags sanity check

### ISSUE THREAD

**Issue by wickedOne** [FriendsOfSymfony/FOSHttpCache#270]
would it make sense to add a sanity check to the [addTag](https://github.com/FriendsOfSymfony/FOSHttpCache/blob/5a3d88ad9093f983601c629ca858943d76216b47/src/ResponseTagger.php#L88) function of the ``ResponseTagger``?

```php
$this->tags = array_merge($this->tags, array_filter($tags, 'strlen'));
```
would prevent an empty tag (``null``, ``''``, ``false``) from being added to the collection of tags.

generation of an empty tag can, for instance, occur if a configured expression function returns an empty string after evaluation

**Comment by dbu** [comment]
good point. you probably would want to know, yes.

in debug mode (e.g. symfony app_dev.php) it definitely would make sense to just throw an exception. for production, a system might prefer small cache issues over aborting page delivery with a 500 page.

maybe the ResponseTagger could get $options constructor argument, with a flag strict_validation that defaults to false. the httpcache symfony bundle can then enable this by default when kernel.debug is true, and offer a configuration option to the user.

**Comment by wickedOne** [comment]
ok, will have a go at that (also next week).

are you suggesting to throw an error when strict validation is enabled and to sanatise when disabled, or do nothing when strict validation is disabled?

a ban like:
```bash
obj.http.X-Host ~ .* && obj.http.X-Url ~ .* && obj.http.content-type ~ .* && obj.http.X-Cache-Tags ~ (mytag-1|)(,.+)?$
``` 
appears to be flushing the entire cache because of the empty value in there (at the end). so even though an emtpy tag is a user error, i'd opt for sanatising in prod mode…

**Comment by dbu** [comment]
tough question. i think this stuff should go into the 2.0 branch because it has potential to change behaviour in subtle ways... there, i would say in strict, we validate tags both when tagging and when invalidating. something like `'' !== (string) $tag` and in non-strict, we simply swallow empty values and neither tag with them nor invalidate them.

**Comment by dbu** [comment]
ups, 2.0 meaning the branch `master`. it is aliased in composer.json to 2.x-dev.

**Comment by wickedOne** [comment]
@dbu could you have a look at [this diff](https://github.com/FriendsOfSymfony/FOSHttpCache/compare/master...wickedOne:tag-sanity?expand=1) and let me know whether i'm on the right track with this one?

**Comment by wickedOne** [comment]
tried working on this yesterday, but run into several issues when installing dependencies:
```bash
  Problem 1
    - Conclusion: don't install php-http/discovery v0.6.4|install puli/manager 1.0.0-beta9
    - Conclusion: don't install puli/manager 1.0.0-beta9
    - php-http/discovery v0.6.3 requires puli/cli ~1.0.0-beta9@dev -> satisfiable by puli/cli[1.0.0-beta10, 1.0.0-beta9, 1.0.x-dev].
    - puli/cli 1.0.0-beta9 requires puli/manager ^1.0-beta9 -> satisfiable by puli/manager[1.0.0-beta10, 1.0.0-beta9, 1.0.x-dev].
    - puli/cli 1.0.0-beta10 requires puli/discovery ^1.0-beta9 -> satisfiable by puli/discovery[1.0.0-beta9, 1.0.x-dev].
    - puli/cli 1.0.x-dev requires puli/discovery ^1.0-beta9 -> satisfiable by puli/discovery[1.0.0-beta9, 1.0.x-dev].
    - puli/manager 1.0.0-beta10 requires puli/discovery ^1.0-beta9 -> satisfiable by puli/discovery[1.0.0-beta9, 1.0.x-dev].
    - puli/manager 1.0.x-dev requires puli/discovery ^1.0-beta9 -> satisfiable by puli/discovery[1.0.0-beta9, 1.0.x-dev].
    - Can only install one of: puli/discovery[1.0.0-beta9, 1.0.0-beta8].
    - Can only install one of: puli/discovery[1.0.x-dev, 1.0.0-beta8].
    - php-http/discovery v0.6.3 requires puli/discovery 1.0.0-beta8 -> satisfiable by puli/discovery[1.0.0-beta8].
    - Installation request for php-http/discovery ^0.6.3 -> satisfiable by php-http/discovery[v0.6.3, v0.6.4].
```
and if i fix ``"php-http/discovery": "0.6.3"`` i get a bunch of config exceptions
```bash
InvalidConfigException: The configuration in /Users/wicliff/Documents/FOSHttpCache/puli.json is invalid:
  packages.[…]: The property env is not defined and the definition does not allow additional properties
```
after which i get
```bash
Running "puli build"

[Puli\ComposerPlugin\PuliRunnerException]
An error occurred while running: /usr/local/php5-5.5.31-20160108-105107/bin/php '[…]FOSHttpCache/vendor/bin/puli' build --no-ansi -vv (status 1): InvalidArgumentException: The flags must be an integer. Got: boolean
```
as i'm completely new to puli: any suggestions (tried php 5.6 as well)?
